File: go/signing.go
package arcusspot

import (
	"crypto/ecdsa"
	"encoding/json"
	"errors"
	"fmt"

	"github.com/ethereum/go-ethereum/common"
	"github.com/ethereum/go-ethereum/common/math"
	"github.com/ethereum/go-ethereum/crypto"
	"github.com/ethereum/go-ethereum/signer/core/apitypes"
)

// TypedDataSigner signs EIP-712 payloads. Implementations must return a
// 65-byte r||s||v signature with v in {27, 28}.
type TypedDataSigner interface {
	// Address returns the signing account address.
	Address() common.Address
	// SignTypedData signs the EIP-712 digest of typedData.
	SignTypedData(typedData Eip712TypedData) ([]byte, error)
}

// PrivateKeySigner signs typed data with an in-memory secp256k1 private key.
type PrivateKeySigner struct {
	key     *ecdsa.PrivateKey
	address common.Address
}

// NewPrivateKeySigner wraps key in a TypedDataSigner.
func NewPrivateKeySigner(key *ecdsa.PrivateKey) *PrivateKeySigner {
	return &PrivateKeySigner{key: key, address: crypto.PubkeyToAddress(key.PublicKey)}
}

// Address implements TypedDataSigner.
func (s *PrivateKeySigner) Address() common.Address { return s.address }

// SignTypedData implements TypedDataSigner.
func (s *PrivateKeySigner) SignTypedData(typedData Eip712TypedData) ([]byte, error) {
	digest, err := HashTypedData(typedData)
	if err != nil {
		return nil, err
	}
	signature, err := crypto.Sign(digest.Bytes(), s.key)
	if err != nil {
		return nil, err
	}
	// crypto.Sign yields v in {0, 1}; wallets and the router expect {27, 28}.
	signature[64] += 27
	return signature, nil
}

// HashTypedData computes the EIP-712 digest of typedData
// (keccak256("\x19\x01" ‖ domainSeparator ‖ hashStruct(message))).
func HashTypedData(typedData Eip712TypedData) (common.Hash, error) {
	converted, err := toAPITypedData(typedData)
	if err != nil {
		return common.Hash{}, err
	}
	digest, _, err := apitypes.TypedDataAndHash(converted)
	if err != nil {
		return common.Hash{}, fmt.Errorf("arcusspot: hash typed data: %w", err)
	}
	return common.BytesToHash(digest), nil
}

// toAPITypedData converts the SDK's wire-shaped typed data to go-ethereum's
// apitypes representation. Like viem, any EIP712Domain entry in Types is
// discarded and rebuilt from the domain fields that are actually present.
func toAPITypedData(typedData Eip712TypedData) (apitypes.TypedData, error) {
	types := apitypes.Types{}
	for name, fields := range typedData.Types {
		if name == "EIP712Domain" {
			continue
		}
		converted := make([]apitypes.Type, len(fields))
		for i, field := range fields {
			converted[i] = apitypes.Type{Name: field.Name, Type: field.Type}
		}
		types[name] = converted
	}

	domain := typedData.Domain
	var domainFields []apitypes.Type
	if domain.Name != "" {
		domainFields = append(domainFields, apitypes.Type{Name: "name", Type: "string"})
	}
	if domain.Version != "" {
		domainFields = append(domainFields, apitypes.Type{Name: "version", Type: "string"})
	}
	if domain.ChainID != nil {
		domainFields = append(domainFields, apitypes.Type{Name: "chainId", Type: "uint256"})
	}
	if domain.VerifyingContract != "" {
		domainFields = append(domainFields, apitypes.Type{Name: "verifyingContract", Type: "address"})
	}
	if domain.Salt != "" {
		domainFields = append(domainFields, apitypes.Type{Name: "salt", Type: "bytes32"})
	}
	types["EIP712Domain"] = domainFields

	apiDomain := apitypes.TypedDataDomain{
		Name:              domain.Name,
		Version:           domain.Version,
		VerifyingContract: domain.VerifyingContract,
		Salt:              domain.Salt,
	}
	if domain.ChainID != nil {
		apiDomain.ChainId = (*math.HexOrDecimal256)(domain.ChainID.Int())
	}

	message, ok := normalizeTypedDataValue(typedData.Message).(map[string]any)
	if !ok {
		return apitypes.TypedData{}, errors.New("arcusspot: typed data message must be an object")
	}

	return apitypes.TypedData{
		Types:       types,
		PrimaryType: typedData.PrimaryType,
		Domain:      apiDomain,
		Message:     message,
	}, nil
}

// normalizeTypedDataValue rewrites json.Number leaves as strings, which
// apitypes accepts for every integer width without float precision loss.
func normalizeTypedDataValue(value any) any {
	switch v := value.(type) {
	case json.Number:
		return v.String()
	case map[string]any:
		out := make(map[string]any, len(v))
		for key, item := range v {
			out[key] = normalizeTypedDataValue(item)
		}
		return out
	case []any:
		out := make([]any, len(v))
		for i, item := range v {
			out[i] = normalizeTypedDataValue(item)
		}
		return out
	default:
		return value
	}
}

// QuoteSigningTask names one EIP-712 payload the taker must sign for a quote.
type QuoteSigningTask struct {
	Venue     Venue
	Kind      string // "approval" | "trade"
	TypedData *Eip712TypedData
}

// GetQuoteSigningTasks lists the signatures a firm quote requires. Every venue
// currently needs exactly one trade signature.
func GetQuoteSigningTasks(quote FirmQuote) []QuoteSigningTask {
	return []QuoteSigningTask{
		{Venue: quote.FirmQuoteVenue(), Kind: "trade", TypedData: quote.TradeTypedData()},
	}
}

// SignQuoteOptions tunes SignQuote.
type SignQuoteOptions struct {
	// Taker overrides the taker address; defaults to the taker bound in the
	// quote's witness, then the signer address.
	Taker common.Address
	// Permits optionally folds EIP-2612 permits into the submit body for a
	// first-time sellToken→Permit2 allowance.
	Permits []Permit
}

// SignQuote signs a firm quote's trade payload and assembles the venue-specific
// submit body. Bebop quotes are not supported.
func SignQuote(quote FirmQuote, signer TypedDataSigner, options *SignQuoteOptions) (SignedQuote, error) {
	if options == nil {
		options = &SignQuoteOptions{}
	}

	taker := options.Taker
	if taker == (common.Address{}) {
		taker = inferTaker(quote)
	}
	if taker == (common.Address{}) {
		taker = signer.Address()
	}
	if taker == (common.Address{}) {
		return nil, errors.New("arcusspot: unable to infer taker address; set SignQuoteOptions.Taker")
	}

	switch q := quote.(type) {
	case *ArcusFirmQuote:
		return signArcusQuote(q, signer, taker, options)
	case *RialtoFirmQuote:
		return signRialtoQuote(q, signer, taker, options)
	case *LifiFirmQuote:
		return signLifiQuote(q, signer, taker, options)
	default:
		return nil, fmt.Errorf("arcusspot: SignQuote does not support venue %s", quote.FirmQuoteVenue())
	}
}

func signArcusQuote(quote *ArcusFirmQuote, signer TypedDataSigner, taker common.Address, options *SignQuoteOptions) (*ArcusSignedQuote, error) {
	signature, err := signer.SignTypedData(quote.ToSign)
	if err != nil {
		return nil, err
	}
	chainID, err := typedDataChainID(&quote.ToSign)
	if err != nil {
		return nil, err
	}
	return &ArcusSignedQuote{
		Venue:       VenueArcus,
		ChainID:     chainID,
		Taker:       taker,
		TypedData:   quote.ToSign,
		Signature:   signature,
		FeePolicyID: quote.Arcus.FeePolicyID,
		Permits:     options.Permits,
	}, nil
}

func signRialtoQuote(quote *RialtoFirmQuote, signer TypedDataSigner, taker common.Address, options *SignQuoteOptions) (*RialtoSignedQuote, error) {
	signature, err := signer.SignTypedData(quote.ToSign)
	if err != nil {
		return nil, err
	}
	chainID, err := typedDataChainID(&quote.ToSign)
	if err != nil {
		return nil, err
	}
	return &RialtoSignedQuote{
		Venue:           VenueRialto,
		ChainID:         chainID,
		Taker:           taker,
		TypedData:       quote.ToSign,
		Signature:       signature,
		Tx:              quote.Tx,
		Permits:         options.Permits,
		QuotedAmountIn:  quote.SellAmount,
		QuotedAmountOut: quote.BuyAmount,
	}, nil
}

func signLifiQuote(quote *LifiFirmQuote, signer TypedDataSigner, taker common.Address, options *SignQuoteOptions) (*LifiSignedQuote, error) {
	signature, err := signer.SignTypedData(quote.ToSign)
	if err != nil {
		return nil, err
	}
	chainID, err := typedDataChainID(&quote.ToSign)
	if err != nil {
		return nil, err
	}
	return &LifiSignedQuote{
		Venue:           VenueLifi,
		ChainID:         chainID,
		Taker:           taker,
		TypedData:       quote.ToSign,
		Signature:       signature,
		Tx:              quote.Tx,
		Raw:             quote.Raw,
		MinBuyAmount:    quote.MinBuyAmount,
		Permits:         options.Permits,
		QuotedAmountIn:  quote.SellAmount,
		QuotedAmountOut: quote.BuyAmount,
	}, nil
}

// SplitSignatureBytes splits a 65-byte r||s||v signature into SwapShell's
// split representation, normalizing v from {0, 1} to {27, 28}.
func SplitSignatureBytes(signature []byte) (SplitSignature, error) {
	if len(signature) != 65 {
		return SplitSignature{}, fmt.Errorf("arcusspot: expected 65-byte signature, got %d bytes", len(signature))
	}
	rawV := signature[64]
	v := rawV
	if v < 27 {
		v += 27
	}
	if v != 27 && v != 28 {
		return SplitSignature{}, fmt.Errorf("arcusspot: unsupported ECDSA recovery byte: %d", rawV)
	}
	return SplitSignature{
		SignatureType: 2,
		V:             v,
		R:             common.BytesToHash(signature[0:32]),
		S:             common.BytesToHash(signature[32:64]),
	}, nil
}

func typedDataChainID(typedData *Eip712TypedData) (uint64, error) {
	if typedData.Domain.ChainID == nil {
		return 0, errors.New("arcusspot: typed data is missing domain.chainId")
	}
	return typedData.Domain.ChainID.Int().Uint64(), nil
}

// inferTaker mirrors the TS SDK: arcus binds the taker in witness.taker, rialto
// in witness.recipient, and other venues in the top-level owner field.
func inferTaker(quote FirmQuote) common.Address {
	message := quote.TradeTypedData().Message
	switch quote.FirmQuoteVenue() {
	case VenueArcus:
		return addressAtPath(message, "witness", "taker")
	case VenueRialto:
		return addressAtPath(message, "witness", "recipient")
	default:
		return addressAtPath(message, "owner")
	}
}

func addressAtPath(message map[string]any, path ...string) common.Address {
	var current any = message
	for _, key := range path {
		object, ok := current.(map[string]any)
		if !ok {
			return common.Address{}
		}
		current, ok = object[key]
		if !ok {
			return common.Address{}
		}
	}
	text, ok := current.(string)
	if !ok || !common.IsHexAddress(text) {
		return common.Address{}
	}
	return common.HexToAddress(text)
}

File: src/signing.ts
import type { Account, Hex, WalletClient } from "viem";
import type {
  ArcusFirmQuote,
  ArcusSignedQuote,
  Eip712TypedData,
  FirmQuote,
  LifiFirmQuote,
  LifiSignedQuote,
  Permit,
  RialtoFirmQuote,
  RialtoSignedQuote,
  SignedQuote,
  SplitSignature,
  Venue,
  ZeroxFirmQuote,
  ZeroxSignedQuote,
} from "./types.js";
import { getZeroxSigningTasks } from "./venues/zerox.js";

export type QuoteSigningTask = {
  venue: FirmQuote["venue"];
  kind: "approval" | "trade";
  typedData: Eip712TypedData;
};

export type SignQuoteOptions = {
  taker?: Hex;
  account?: Account | Hex;
  /**
   * Optional EIP-2612 permits to fold into SwapShell.permits[] (arcus). Use when
   * the taker's sellToken→Permit2 allowance is missing and they sign a one-time
   * permit instead of an approve() tx
   */
  permits?: Permit[];
  /**
   * Arcus only. Echo the `/quote` `builderFeeBps` onto the submit body so
   * settlement matches the quoted fee plan.
   */
  builderFeeBps?: number;
};

export function getQuoteSigningTasks(quote: FirmQuote): QuoteSigningTask[] {
  if (quote.venue === "zerox") return getZeroxSigningTasks(quote);
  return [{ venue: quote.venue, kind: "trade", typedData: tradeTypedData(quote) }];
}

export async function signQuote(
  quote: FirmQuote,
  walletClient: WalletClient,
  options: SignQuoteOptions = {},
): Promise<SignedQuote> {
  const taker = options.taker ?? inferTaker(quote) ?? walletClient.account?.address;
  if (!taker) {
    throw new Error("Unable to infer taker address; pass signQuote(..., { taker })");
  }

  if (quote.venue === "zerox") {
    return signZeroxQuote(quote, walletClient, taker, options.account);
  }
  if (quote.venue === "arcus") {
    return signArcusQuote(quote, walletClient, taker, options);
  }
  if (quote.venue === "rialto") {
    return signRialtoQuote(quote, walletClient, taker, options);
  }
  if (quote.venue === "lifi") {
    return signLifiQuote(quote, walletClient, taker, options);
  }
  throw new Error(`signQuote does not support venue ${quote.venue}`);
}

export async function signTypedDataWithViem(
  walletClient: WalletClient,
  typedData: Eip712TypedData,
  account?: Account | Hex,
): Promise<Hex> {
  const signingAccount = account ?? walletClient.account ?? undefined;
  return walletClient.signTypedData({
    account: signingAccount,
    domain: typedData.domain,
    types: stripEip712Domain(typedData.types),
    primaryType: typedData.primaryType,
    message: typedData.message,
  } as Parameters<WalletClient["signTypedData"]>[0]);
}

export function splitSignature(signature: Hex): SplitSignature {
  const hex = signature.slice(2);
  if (hex.length !== 130) {
    throw new Error(`Expected 65-byte signature, got ${hex.length / 2} bytes`);
  }

  const rawV = Number.parseInt(hex.slice(128, 130), 16);
  const v = rawV < 27 ? rawV + 27 : rawV;
  if (v !== 27 && v !== 28) {
    throw new Error(`Unsupported ECDSA recovery byte: ${rawV}`);
  }

  return {
    signatureType: 2,
    r: `0x${hex.slice(0, 64)}`,
    s: `0x${hex.slice(64, 128)}`,
    v,
  };
}

async function signZeroxQuote(
  quote: ZeroxFirmQuote,
  walletClient: WalletClient,
  taker: Hex,
  account?: Account | Hex,
): Promise<ZeroxSignedQuote> {
  const signature = await signTypedDataWithViem(walletClient, tradeTypedData(quote), account);

  // EIP-2612 Permit
  const permits = quote.approval
    ? [
        permitFromApproval(
          await signTypedDataWithViem(walletClient, quote.approval.eip712, account),
          quote.approval.eip712,
        ),
      ]
    : undefined;

  return {
    venue: "zerox",
    chainId: typedDataChainId(tradeTypedData(quote)),
    taker,
    typedData: tradeTypedData(quote),
    signature,
    quotedAmountIn: quote.sellAmount,
    quotedAmountOut: quote.buyAmount,
    ...(permits ? { permits } : {}),
  };
}

async function signArcusQuote(
  quote: ArcusFirmQuote,
  walletClient: WalletClient,
  taker: Hex,
  options: SignQuoteOptions,
): Promise<ArcusSignedQuote> {
  const signature = await signTypedDataWithViem(walletClient, quote.toSign, options.account);

  const permits = options.permits?.length ? options.permits : undefined;

  return {
    venue: "arcus",
    chainId: typedDataChainId(quote.toSign),
    taker,
    typedData: quote.toSign,
    signature,
    ...(permits ? { permits } : {}),
    ...(options.builderFeeBps != null ? { builderFeeBps: options.builderFeeBps } : {}),
  };
}

async function signRialtoQuote(
  quote: RialtoFirmQuote,
  walletClient: WalletClient,
  taker: Hex,
  options: SignQuoteOptions,
): Promise<RialtoSignedQuote> {
  const signature = await signTypedDataWithViem(walletClient, quote.toSign, options.account);

  const permits = options.permits?.length ? options.permits : undefined;

  return {
    venue: "rialto",
    chainId: typedDataChainId(quote.toSign),
    taker,
    typedData: quote.toSign,
    signature,
    tx: quote.tx,
    quotedAmountIn: quote.sellAmount,
    quotedAmountOut: quote.buyAmount,
    ...(permits ? { permits } : {}),
  };
}

async function signLifiQuote(
  quote: LifiFirmQuote,
  walletClient: WalletClient,
  taker: Hex,
  options: SignQuoteOptions,
): Promise<LifiSignedQuote> {
  const signature = await signTypedDataWithViem(walletClient, quote.toSign, options.account);

  const permits = options.permits?.length ? options.permits : undefined;

  return {
    venue: "lifi",
    chainId: typedDataChainId(quote.toSign),
    taker,
    typedData: quote.toSign,
    signature,
    tx: quote.tx,
    raw: quote.raw,
    minBuyAmount: quote.minBuyAmount,
    quotedAmountIn: quote.sellAmount,
    quotedAmountOut: quote.buyAmount,
    ...(permits ? { permits } : {}),
  };
}

/**
 * Convert an EIP-2612 approval signature + its typed data into a Permit
 * entry suitable for SwapShell.permits[]. Reads token / value / deadline
 * from the same typed data the user just signed, so the on-chain call can
 * recover the same digest.
 */
function permitFromApproval(signature: Hex, typedData: Eip712TypedData): Permit {
  const hex = signature.slice(2);
  if (hex.length !== 130) {
    throw new Error(`Expected 65-byte signature, got ${hex.length / 2} bytes`);
  }
  const rawV = Number.parseInt(hex.slice(128, 130), 16);
  const v = rawV < 27 ? rawV + 27 : rawV;
  if (v !== 27 && v !== 28) {
    throw new Error(`Unsupported ECDSA recovery byte: ${rawV}`);
  }

  const token = typedData.domain.verifyingContract;
  const value = typedData.message.value;
  const deadline = typedData.message.deadline;
  if (typeof token !== "string" || !/^0x[0-9a-fA-F]{40}$/.test(token)) {
    throw new Error("EIP-2612 approval missing verifyingContract (token address)");
  }
  if (value == null || deadline == null) {
    throw new Error("EIP-2612 approval message missing value/deadline");
  }

  return {
    token: token as Hex,
    value: String(value),
    deadline: String(deadline),
    v,
    r: `0x${hex.slice(0, 64)}`,
    s: `0x${hex.slice(64, 128)}`,
  };
}

function stripEip712Domain(types: Eip712TypedData["types"]): Eip712TypedData["types"] {
  const { EIP712Domain: _domain, ...rest } = types;
  return rest;
}

function inferTaker(quote: FirmQuote): Hex | undefined {
  if (quote.venue === "arcus") {
    const witnessTaker = nestedValue(quote.toSign.message, ["witness", "taker"]);
    return isHexAddress(witnessTaker) ? witnessTaker : undefined;
  }
  if (quote.venue === "rialto") {
    // RialtoSwap witness recipient is the taker (funds land there).
    const recipient = quote.toSign.message.witness.recipient;
    return isHexAddress(recipient) ? recipient : undefined;
  }

  const approvalOwner = quote.venue === "zerox" ? quote.approval?.eip712.message.owner : undefined;
  if (isHexAddress(approvalOwner)) return approvalOwner;

  const typedData = maybeTradeTypedData(quote);
  const tradeOwner = typedData?.message.owner;
  if (isHexAddress(tradeOwner)) return tradeOwner;

  if (!typedData) return undefined;
  const recipient = nestedValue(typedData.message, ["slippageAndActions", "recipient"]);
  return isHexAddress(recipient) ? recipient : undefined;
}

function tradeTypedData(quote: FirmQuote): Eip712TypedData {
  const typedData = maybeTradeTypedData(quote);
  if (!typedData) {
    throw new Error(`Quote is missing trade typed data for venue ${quote.venue}`);
  }
  return typedData;
}

function maybeTradeTypedData(quote: FirmQuote): Eip712TypedData | undefined {
  return quote.venue === "zerox" ? (quote.toSign ?? quote.trade?.eip712) : quote.toSign;
}

function typedDataChainId(typedData: Eip712TypedData): number {
  const chainId = typedData.domain.chainId;
  if (typeof chainId === "number") return chainId;
  if (typeof chainId === "bigint") return Number(chainId);
  if (typeof chainId === "string") return Number(chainId);
  throw new Error("Typed data is missing domain.chainId");
}

function nestedValue(source: Record<string, unknown>, path: string[]): unknown {
  let current: unknown = source;
  for (const key of path) {
    if (!current || typeof current !== "object" || !(key in current)) return undefined;
    current = (current as Record<string, unknown>)[key];
  }
  return current;
}

function isHexAddress(value: unknown): value is Hex {
  return typeof value === "string" && /^0x[0-9a-fA-F]{40}$/.test(value);
}
